### DISCUSSION TITLE
Expressing a dependency for a custom_target

### DISCUSSION THREAD

**Discussion by medoc92** [mesonbuild/meson#13320] [Q&A] [answered]
Hi,
I am using meson/ninja to build a desktop search program, Recoll.
The application is made of a library, command-line programs, and a GUI.
The library and command line programs are built the normal way.
The GUI is built by Qt qmake for now, run from meson by using a custom_target command to run qmake and make. This needs the library.

On some machines with many cores, it seems that the GUI build will sometimes start before the library is finished building, causing a build error. Using a -j argument to ninja with a lower value will fix the error.

It seems that a custom_target does not have a dependencies attribute. How can I tell meson that the target depends on the library build ?
 
Refs to the meson.build files if this is useful:

Main: https://framagit.org/medoc92/recoll/-/blob/master/src/meson.build?ref_type=heads

GUI: https://framagit.org/medoc92/recoll/-/blob/master/src/qtgui/meson.build?ref_type=heads

**Comment by nils-werner** [comment, marked as answer]
[Of course you can](https://mesonbuild.com/Reference-manual_functions.html#custom_target_depends):

```
custom_target(
    ...
    depends : my_other_target,
)
```

**Comment by medoc92** [comment]
Thanks a lot ! I had tried with 'dependencies', which did not work, and did not go all the way to the reference manual. I do feel stupid...

**Comment by eli-schwartz** [comment]
> The GUI is built by Qt qmake for now, run from meson by using a custom_target command to run qmake and make. This needs the library.

As a matter of curiosity -- is there any specific issue preventing from building the GUI using meson's builtin Qt support?

**Comment by medoc92** [comment]
The motivation for the switch to meson was to replace the autotools which were used to build the library and the command line utilities. The GUI is built with qmake/make on all platforms (Linux/Windows/MacOS), and not touching this allowed reaching the main goal faster. Maybe the GUI build will also switch in the future, but the good integration of qcreator/qmake with both Qt and the different platform tools and compilers makes it difficult to beat for simplicity.

**Reply by eli-schwartz** [reply]
@jeandet added a meson plugin to qtcreator, although I haven't tried it myself.

Thanks for the context.

**Reply by jeandet** [reply]
@medoc92, the limitations on top of my head for QTC Meson plugin together with Meson itself regarding Qt SDK integration are:
- No out of the box PySide/Shiboken support (but easy to hack around)
- No QTC locator integration yet (could be handy to issue Meson commands)
- QTC can't modify Meson build files, so adding new files has to be done manually (should be possible in the future)
- QML support not there yet #13304

Those are pretty minor I would say, I personally use QTC with Meson daily for all my C++ projects either with Qt or not. So far, my experience is that switching to Meson makes multi-platform simpler most of the time and even cross compilation. 
Note that WrapDB has quite a good list of C++ dependencies that makes Windows build for example way simpler.

**Comment by medoc92** [comment]
@jeandet Thanks a lot for this information. Not changing the GUI build satisfied the requirement to ditch the autotools and was the minimum effort. I'll have a look at using meson for everything when I get a little more time.
